File: c++/9_rolling_average.cpp
// Theory:
// NumPy's core advantage over plain Python loops is vectorization:
// operations run as compiled C loops over contiguous memory, avoiding
// per-iteration interpreter overhead. C++ is already compiled, so the
// analogous win here isn't interpreter-overhead avoidance - it's an
// algorithmic one: a naive rolling-window computation recomputes the sum
// of each window from scratch, costing O(n*w) time for n elements and
// window size w. The fix is the same prefix-sum (cumulative sum) trick
// NumPy's cumsum-based approach exploits: precompute cumulative sums once
// in O(n), then each window's sum is just a difference of two prefix-sum
// entries, dropping the total cost to O(n) regardless of window size -
// std::partial_sum is the STL's direct equivalent of numpy.cumsum.

// Headers & STL components:
// <numeric> - std::partial_sum (prefix sums), std::accumulate (a window sum directly)
// <vector> - the input/output container type here
// <algorithm> - std::transform, elementwise vectorized-style operations

// Task:
// Given a vector of numbers prices and a window size w, return the simple
// moving average over each full window of size w - a vector of length
// prices.size() - w + 1. Assume 1 <= w <= prices.size().
// Test1: Inputs: prices = {1, 2, 3, 4, 5}, w = 2   Outputs: {1.5, 2.5, 3.5, 4.5}
// Test2: Inputs: prices = {1, 2, 3, 4, 5}, w = 3   Outputs: {2.0, 3.0, 4.0}
// Test3: Inputs: prices = {5, 5, 5, 5}, w = 2   Outputs: {5.0, 5.0, 5.0}
// Hint:
// Ugly way: nested loop - sum each window of w elements from scratch.
// Right way: std::partial_sum - precompute prefix sums once, then each
// window's sum is a single subtraction.

// Solutions:
#include <cassert>
#include <iostream>
#include <numeric>
#include <vector>

// Space - Time Complexity analysis:
// space: O(n) - result vector holds n-w+1 averages
// time: O(n*w) - n windows, each summed from scratch in O(w) via accumulate

std::vector<double> rolling_average_ugly(const std::vector<double>& prices, int w) {
    size_t window = static_cast<size_t>(w);  // O(1) | O(1)
    std::vector<double> result;  // O(1) | O(1)
    for (size_t i = 0; i + window <= prices.size(); ++i) {  // O(1) | O(n)
        double window_sum = std::accumulate(prices.begin() + i, prices.begin() + i + window, 0.0);  // O(w) | O(w)
        result.push_back(window_sum / w);  // O(n) | O(1)
    }
    return result;  // O(1) | O(1)
}

// Space - Time Complexity analysis:
// space: O(n) - the prefix-sum vector and the result vector both scale with n
// time: O(n) - two separate linear passes: build prefix sums, then one O(1) subtraction per window

std::vector<double> rolling_average(const std::vector<double>& prices, int w) {
    // prefix[i] = sum of the first i prices; prefix[0] = 0
    std::vector<double> prefix(prices.size() + 1, 0.0);  // O(n) | O(n)
    // build the cumulative sums in one pass
    std::partial_sum(prices.begin(), prices.end(), prefix.begin() + 1);  // O(1) | O(n)
    size_t window = static_cast<size_t>(w);  // O(1) | O(1)
    std::vector<double> result;  // O(1) | O(1)
    // each window's sum is just the difference of two prefix entries
    for (size_t i = 0; i + window <= prices.size(); ++i) {  // O(1) | O(n)
        double window_sum = prefix[i + window] - prefix[i];  // O(1) | O(1)
        result.push_back(window_sum / w);  // O(n) | O(1)
    }
    return result;  // O(1) | O(1)
}

// Tests:
void validate() {
    struct Case {
        std::vector<double> prices;
        int w;
        std::vector<double> expected;
    };
    std::vector<Case> cases = {
        {{1, 2, 3, 4, 5}, 2, {1.5, 2.5, 3.5, 4.5}},
        {{1, 2, 3, 4, 5}, 3, {2.0, 3.0, 4.0}},
        {{5, 5, 5, 5}, 2, {5.0, 5.0, 5.0}},
        {{2, 4, 6}, 3, {4.0}},
    };
    for (const auto& c : cases) {
        assert(rolling_average_ugly(c.prices, c.w) == c.expected);
        assert(rolling_average(c.prices, c.w) == c.expected);
    }
    std::cout << "SUCCESS" << std::endl;
}

int main() {
    validate();
    return 0;
}

File: python/1_core/9_rolling_average.py
# Theory:
## NumPy's core advantage over plain Python loops is vectorization:
## operations like sums, means, and elementwise arithmetic run as compiled
## C loops over contiguous memory, avoiding Python's per-iteration
## interpreter overhead - often 10-100x faster for numeric workloads. A
## naive rolling-window computation (like a moving average) recomputes the
## sum of each window from scratch, costing O(n*w) time for n elements and
## window size w. The vectorized fix uses a prefix-sum (cumulative sum)
## trick: precompute cumulative sums once in O(n), then each window's sum
## is just a difference of two prefix-sum entries, dropping the total cost
## to O(n) regardless of window size - the same algorithmic idea NumPy's
## cumsum-based tricks (and pandas' .rolling()) exploit under the hood.


# Packages:
## numpy.cumsum - cumulative sum, the basis of an O(n) rolling-window trick
## numpy.convolve - sliding-window sum/correlation in one vectorized call
## pandas.Series.rolling - built-in rolling-window aggregations (mean, sum, std, ...)
## numpy.mean / numpy.array - vectorized elementwise operations over arrays


# Task:
## Given a list of numbers prices and a window size w, return the simple
## moving average over each full window of size w - a list of length
## len(prices) - w + 1. Assume 1 <= w <= len(prices).
## Test1: Inputs: prices = [1, 2, 3, 4, 5], w = 2   Outputs: [1.5, 2.5, 3.5, 4.5]
## Test2: Inputs: prices = [1, 2, 3, 4, 5], w = 3   Outputs: [2.0, 3.0, 4.0]
## Test3: Inputs: prices = [5, 5, 5, 5], w = 2   Outputs: [5.0, 5.0, 5.0]
## Hint:
## Ugly way: nested loop - sum each window of w elements from scratch.
## Right way: prefix sums (or numpy.cumsum) - precompute cumulative sums
## once, then each window's sum is a single subtraction.


# Solutions:
import numpy as np


# Space - Time Complexity analysis:
## space: O(n) - result list holds n-w+1 averages
## time: O(n*w) - n windows, each summed from scratch in O(w) via slicing

def rolling_average_ugly(prices, w):
    result = []  # O(1) | O(1)
    for i in range(len(prices) - w + 1):  # O(1) | O(n)
        window_sum = sum(prices[i:i + w])  # O(w) | O(w)
        result.append(window_sum / w)  # O(n) | O(1)
    return result  # O(1) | O(1)


# Space - Time Complexity analysis:
## space: O(n) - the numpy array, prefix-sum array, and result all scale with n
## time: O(n) - np.cumsum is a single vectorized pass; the window-sum subtraction is vectorized over all windows at once

def rolling_average(prices, w):
    # numpy array so cumsum runs as a vectorized C loop, not a Python loop
    arr = np.array(prices, dtype=float)  # O(n) | O(n)
    # prefix[i] = sum of the first i prices; a leading 0 makes prefix[0] = 0
    prefix = np.cumsum(np.insert(arr, 0, 0))  # O(n) | O(n)
    # every window's sum at once: prefix[w:] - prefix[:-w], no explicit loop over windows
    window_sums = prefix[w:] - prefix[:-w]  # O(n) | O(n)
    return (window_sums / w).tolist()  # O(n) | O(n)


# Tests:
def validate():
    cases = [
        ([1, 2, 3, 4, 5], 2, [1.5, 2.5, 3.5, 4.5]),
        ([1, 2, 3, 4, 5], 3, [2.0, 3.0, 4.0]),
        ([5, 5, 5, 5], 2, [5.0, 5.0, 5.0]),
        ([2, 4, 6], 3, [4.0]),
    ]
    for fn in (rolling_average_ugly, rolling_average):
        for prices, w, expected in cases:
            result = fn(prices, w)
            assert result == expected, (
                f"{fn.__name__}({prices!r}, {w}): expected {expected}, got {result}"
            )
    print("SUCCESS")


if __name__ == "__main__":
    validate()
